src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs

```diff
@@ -173,25 +232,18 @@ public override void VisitBinaryExpression(BinaryExpressionSyntax node)
 
         public override void VisitDeclarationExpression(DeclarationExpressionSyntax node)
         {
-            var context = node?.Parent  // ArgumentSyntax
-                              ?.Parent  // ArgumentListSyntax
-                              ?.Parent; // invocation/constructor initializer
-            switch (context?.Kind())
-            {
-                case SyntaxKind.InvocationExpression:
-                case SyntaxKind.ObjectCreationExpression:
-                case SyntaxKind.ThisConstructorInitializer:
-                case SyntaxKind.BaseConstructorInitializer:
-                    var local = SourceLocalSymbol.MakeOutVariable(_scopeBinder.ContainingMemberOrLambda, _scopeBinder, _enclosingBinderOpt, node.Type(), node.Identifier(), context);
-                    _localsBuilder.Add(local);
-                    break;
-                default:
-
-                    // It looks like we are deling with a syntax tree that has a shape that could never be
-                    // produced by the LanguageParser, including all error conditions. 
-                    // Out Variable declarations can only appear in an argument list of the syntax nodes mentioned above.
-                    throw ExceptionUtilities.UnexpectedValue(context?.Kind());
-            }
+            var argumentSyntax = (ArgumentSyntax)node?.Parent;
+            var argumentListSyntax = (BaseArgumentListSyntax)argumentSyntax?.Parent;
+            var invocationSyntax = argumentListSyntax?.Parent;
+            _localsBuilder.Add(SourceLocalSymbol.MakeLocalSymbolWithEnclosingContext(
+                containingSymbol: _scopeBinder.ContainingMemberOrLambda,
+                scopeBinder: _scopeBinder,
+                nodeBinder: _enclosingBinder,
+                typeSyntax: node.Type(),
+                identifierToken: node.Identifier(),
+                kind: LocalDeclarationKind.RegularVariable,
+                nodeToBind: _nodeToBind,
+                forbiddenZone: invocationSyntax));
```

Comment:
That would have the same behavior. We check for use-before-declaration first, so either way would produce the same diagnostics.

---
In reply to: [76324205](https://github.com/dotnet/roslyn/pull/13274#discussion_r76324205) [](ancestors = 76324205)